pkg/sql/schemachanger/scplan/scstage/build.go

```diff
@@ -21,249 +21,420 @@ import (
-func makeBuildContext(init scpb.State, phase scop.Phase, g *scgraph.Graph) buildCtx {
-	// Fetch the ranks of the nodes, which will be used to evaluate edges
-	// in topological order.
-	nodeRanks, err := g.GetNodeRanks()
-	if err != nil {
-		panic(err)
+// makeStageBuilder returns a stage builder with an operation type for which
+// progress can be made. Defaults to the mutation type if none make progress.
+func (b buildState) makeStageBuilder() (sb stageBuilder) {
+	opTypes := []scop.Type{scop.BackfillType, scop.ValidationType, scop.MutationType}
+	switch b.phase {
+	case scop.StatementPhase, scop.PreCommitPhase:
+		// We don't allow expensive operations pre-commit.
+		opTypes = []scop.Type{scop.MutationType}
 	}
-	return buildCtx{
-		g:                     g,
-		phase:                 phase,
-		isRevertiblePreferred: true,
-		nodeRanks:             nodeRanks,
-		// TODO(ajwerner): deal with the case where the target status was
-		// fulfilled by something that preceded the initial state.
-		state:     init,
-		fulfilled: make(map[*scpb.Node]struct{}, len(nodeRanks)),
+	for _, opType := range opTypes {
+		sb = b.makeStageBuilderForType(opType)
+		if sb.canMakeProgress() {
+			break
+		}
 	}
+	return sb
 }
 
-// tryBuildStage tries to build a stage using the provided op-edges.
-// This is done on a best-effort basis, if not possible nil is returned.
-func (b buildCtx) tryBuildStage(edges []*scgraph.OpEdge) *Stage {
-	// Group the op edges a per-type basis.
-	opTypes := make(map[scop.Type][]*scgraph.OpEdge)
-	for _, oe := range edges {
-		opTypes[oe.Type()] = append(opTypes[oe.Type()], oe)
+// makeStageBuilderForType creates and populates a stage builder for the given
+// op type.
+func (b buildState) makeStageBuilderForType(opType scop.Type) stageBuilder {
+	sb := stageBuilder{
+		bs:         b,
+		opType:     opType,
+		current:    make([]currentTargetState, len(b.state.Nodes)),
+		fulfilling: map[*scpb.Node]struct{}{},
+	}
+	for i, n := range b.state.Nodes {
+		t := sb.makeCurrentTargetState(n)
+		sb.current[i] = t
 	}
+	// Greedily try to make progress by going down op edges when possible.
+	for isDone := false; !isDone; {
+		isDone = true
+		for i, t := range sb.current {
+			if t.e == nil {
+				continue
+			}
+			if sb.hasUnmetInboundDeps(t.e.To()) {
+				continue
+			}
+			if sb.hasUnmeetableOutboundDeps(t.e.To()) {
+				continue
+			}
+			sb.opEdges = append(sb.opEdges, t.e)
+			sb.fulfilling[t.e.To()] = struct{}{}
+			sb.current[i] = sb.nextTargetState(t)
+			isDone = false
+		}
+	}
+	return sb
+}
+
+// stageBuilder contains the state for building one stage.
+type stageBuilder struct {
+	bs         buildState
+	opType     scop.Type
+	current    []currentTargetState
+	fulfilling map[*scpb.Node]struct{}
+	opEdges    []*scgraph.OpEdge
+}
+
+type currentTargetState struct {
+	n *scpb.Node
+	e *scgraph.OpEdge
+
+	// opFlag is true iff this stage already includes an op edge with ops for this
+	// target.
+	opFlag bool
```

Comment:
can we call this something more verbose? I find myself having to think a bit every time I read it `alreadyEmittedOpsForCurrentStage`?